### ISSUE TITLE
Infant pipeline troubleshooting

### ISSUE THREAD

**Issue by TeddyTuresky** [yeatmanlab/pyAFQ#966]
Hi all,

I'm trying to run pyAFQ on an infant participant using the example here: https://github.com/yeatmanlab/pyAFQ/blob/master/examples/tutorial_examples/plot_baby_afq.py (thank you for making me aware of this, John!). 

The pipeline generates *.trk and *.json cleaned bundle files for all the tracts in the recent Grotheer et al. (2023) preprint. However, it seems to run into some problems shortly after, when generating the *desc-profiles_dwi.csv file. All rows, regardless of tract, look similar to the example rows below:

,tractID,nodeID,dti_fa,dti_md
0,ARC_L,0,0.0,9.99999997475243e-07
1,ARC_L,1,0.0,9.99999997475242e-07
2,ARC_L,2,0.0,9.99999997475243e-07
3,ARC_L,3,0.0,9.99999997475242e-07
4,ARC_L,4,0.0,9.999999974752423e-07
5,ARC_L,5,0.0,9.999999974752427e-07
6,ARC_L,6,0.0,9.999999974752425e-07
7,ARC_L,7,0.0,9.99999997475243e-07

The pipeline then encounters an error when running the plotly viz scripts (error messages pasted below).

From what I can tell, the *.trk files look great, with the exception of the two corpus callosum bundles, one of which had no interhemispheric fibers (FP) and the other had no fibers at all (FA). I'm not sure whether this is a related problem. The contents of the *slCount_dwi.csv file are pasted below:

,n_streamlines,n_streamlines_clean
ARC_L,753,200
ARC_R,342,153
ATR_L,2120,1205
ATR_R,2181,981
CGC_L,1432,689
CGC_R,2671,1318
CST_L,1583,1116
CST_R,2659,1538
FA,0,0
FP,1554,722
IFO_L,103,64
IFO_R,86,63
ILF_L,3301,1544
ILF_R,3699,2037
MdLF_L,398,306
MdLF_R,771,455
SLF_L,6085,4097
SLF_R,10615,5679
UNC_L,2824,1521
UNC_R,2738,1244
whole_brain,45915,24932

Any guidance on resolving some of these issues be greatly appreciated!

Many thanks,
Teddy

Relevant error messages:
INFO:AFQ:Generating colorful lines from tractography...
/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/viz/plotly_backend.py:105: RuntimeWarning:

divide by zero encountered in double_scalars

/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/numpy/core/numeric.py:942: RuntimeWarning:

invalid value encountered in multiply

/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/viz/plotly_backend.py:191: RuntimeWarning:

invalid value encountered in multiply

Traceback (most recent call last):
  File "/n/intermediate_path/Lab/pyafq-pract/pyafq-inf-simple.py", line 38, in <module>
    my_afq.export_all()
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/api/group.py", line 482, in export_all
    export_all_helper(self, seg_algo, xforms, indiv, viz)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/api/utils.py", line 140, in export_all_helper
    api_afq_object.export("all_bundles_figure")
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/api/group.py", line 431, in export
    results[subject][session] = wf_dict[attr_name]
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/pimms/calculation.py", line 470, in __getitem__
    self._run_node(self.plan.efferents[k])
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/pimms/calculation.py", line 534, in _run_node
    if not found: res = node(self)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/pimms/calculation.py", line 91, in __call__
    result = self.function(*args)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/tasks/viz.py", line 107, in viz_bundles
    figure = viz_backend.visualize_bundles(
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/viz/plotly_backend.py", line 363, in visualize_bundles
    _plot_profiles(
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/AFQ/viz/plotly_backend.py", line 200, in _plot_profiles
    go.Scatter3d(
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/graph_objs/_scatter3d.py", line 2590, in __init__
    self["line"] = _v
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/basedatatypes.py", line 4811, in __setitem__
    self._set_compound_prop(prop, value)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/basedatatypes.py", line 5222, in _set_compound_prop
    val = validator.validate_coerce(val, skip_invalid=self._skip_invalid)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/_plotly_utils/basevalidators.py", line 2467, in validate_coerce
    v = self.data_class(v, skip_invalid=skip_invalid, _validate=_validate)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/graph_objs/scatter3d/_line.py", line 901, in __init__
    self["color"] = _v
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/basedatatypes.py", line 4819, in __setitem__
    self._set_prop(prop, value)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/basedatatypes.py", line 5163, in _set_prop
    raise err
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/plotly/basedatatypes.py", line 5158, in _set_prop
    val = validator.validate_coerce(val)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/_plotly_utils/basevalidators.py", line 1375, in validate_coerce
    self.raise_invalid_elements(invalid_els)
  File "/n/home_fasse/tturesky/.local/lib/python3.9/site-packages/_plotly_utils/basevalidators.py", line 305, in raise_invalid_elements
    raise ValueError(
ValueError: 
    Invalid element(s) received for the 'color' property of scatter3d.line
        Invalid elements include: ['rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)', 'rgba(nan, nan, nan, 0.999)']

**Comment by 36000** [comment]
I think the visualization error is likely a result of the dti_fa / dti_md being wrong. One reason the FA/MD might be 0 is a bad brain mask, in this case a brain mask which is empty. If you load the brain mask, does it look reasonable? another thing to look at is the FA/MD files generated by the pipeline. Do they look reasonable? If you delete the FA/MD files generated by pyAFQ and re-run the pipeline, they will regenerate automatically. 

For the bundles, my guess is the DIPY tractography is having trouble in infant brains. You could try reducing stop_threshold, but that might make the other bundles less clean. I will ask Mareike Grotheer.

**Comment by arokem** [comment]
Tagging in @grotheer with her GitHub handle, in case she has ideas about the missing bundles. IIRC, the callosal bundles in particular are exquisitely sensitive to the definition of the WM mask, so I'd double check the WM mask around the Corpus Callosum.

**Comment by 36000** [comment]
Another option is to use Mrtrix, as described in this repo: https://github.com/EduNeuroLab/wmDevAroundBirth
In the mrtrix_pipeline.ipynb file

**Comment by TeddyTuresky** [comment]
Thank you both for all the guidance. All the preprocessing and tractography were done in MRtrix using a pipeline similar to the one in the EduNeuroLab repository (I hope). In the past, using ACT with poor tissue segmentations (to Ariel's point, mostly underestimation of white matter) led to poor reconstructions in several tracts, but I'm now using tissue segmentations generated by a combination of iBEATv2, FreeSurfer, and Infant FreeSurfer, which seems to have greatly restored tract reconstructions except for the callosal bundles. It would certainly make sense if white matter were still being underestimated in these regions; I mainly mentioned this issue in case it was related to the tract profile estimation issue, but it sounds as though it's probably not. 

The brain and white matter masks (from the 5tt file and with the anterior callosum ROIs in subject space), as well as the FA and MD files, look reasonable from what I can tell, except for maybe the midline callosum ROI (i.e., not sure why it's so large)...what do you think though?
<img width="888" alt="brainmask" src="https://github.com/yeatmanlab/pyAFQ/assets/28765464/3abe3e06-895d-4687-aa20-f08497f7a3f8">
<img width="890" alt="wm" src="https://github.com/yeatmanlab/pyAFQ/assets/28765464/bd00120d-0103-443a-a19d-b869a4e22dab">
<img width="830" alt="fa" src="https://github.com/yeatmanlab/pyAFQ/assets/28765464/3d364573-5481-438c-9d97-b82e5ef0c581">
<img width="833" alt="md" src="https://github.com/yeatmanlab/pyAFQ/assets/28765464/93797ed2-9043-4f30-a2a8-f1ae42b088bf">

**Comment by 36000** [comment]
The bug is very weird then. are the FA/MD files from outside pyAFQ by any chance?

**Comment by TeddyTuresky** [comment]
Welp, in drafting a response, I stumbled onto a couple mistakes I made...first, I think you were right to suspect the FA and MD files. The above screenshots for the masks correspond to the infant I was testing, but the FA and MD screenshots correspond to a different, older participant (on whom the standard pyAFQ pipeline ran fine). 🤦🏻‍♂️ In actuality, the infant FA image was empty and the infant MD image looked like the brain mask. 
My second mistake, which I'm assuming explains the FA and MD images, was that I had kept the bval limits from the baby_pyAFQ example in the GroupAFQ object I was using, thinking that our non-b0 bvals were all exactly 1000. But it turns out they were _roughly_ 1000. Once I removed these parameters, the pipeline finished without issue (minus the unrelated problem with the callosum bundles). The tract profile csv now contains FA and MD values that look reasonable and the viz folders have been populated! 
In any event, thank you again for all the help getting this running! ...and, of course, for it in general.
